### DISCUSSION TITLE
Netbeans IDE extension

### DISCUSSION THREAD

**Discussion by v-nchatitai** [community/community#18405] [Copilot Conversations]
Does Copilot have a Netbeans IDE extenstion? It's my preferred IDE.

**Comment by Zerotask** [comment]


**Reply by Chris2011** [reply]
It is still as not as powerful as VS Code or WebStorm but still usable nowadays with an acceptable range of developers to think about creating a plugin. Also Copilot is available for Vim and for IntelliJ so this is a valid point to have this feature for other Editors/IDEs.

The problem is, if the developers of copilot doesn't want to reuse the code from the IntelliJ plugin, change the IDE specific code and use it for NetBeans, we as a community need a documentation to know how to implement it into other Editors/IDE. But there isn't anything afaik.

**Reply by gorlovka** [reply]
Netbeans is great

**Comment by RobbiewOnline** [comment]
> Does Copilot have a Netbeans IDE extenstion? It's my preferred IDE.

I'd use Co-pilot if there was a NetBeans plugin for Java Development.  I'm not going to starting using Visual Studio Code (tried it and the Java / Gradle / Tomcat plugins keep giving me grief, so I'd rather stick with NetBeans).

I plan to use Co-pilot in Visual Studio For Mac for my C# development.

**Comment by irfina** [comment]
Yes, copilot plugin for NetBeans please

**Comment by diegofalcinelli** [comment]
Yes, copilot plugin for NetBeans please

**Comment by shanjitsingh1** [comment]
**Yes, copilot plugin for NetBeans please**

**Reply by hunterbit** [reply]
Yes, copilot plugin for NetBeans please

**Comment by atilafreitasjr** [comment]
Yes, copilot plugin for NetBeans please

**Comment by Montego02** [comment]
Yes, copilot plugin for NetBeans please

**Comment by Chris2011** [comment]
Can you please just stop spamming? I think they already got it.

**Comment by jdcastro** [comment]
Yes, copilot plugin for NetBeans please

**Comment by Anas-Hammou** [comment]
Yes, copilot plugin for NetBeans please

**Comment by javactive** [comment]
It would be great to have it for NetBeans, I don't get why still doesn't have it. 
NetBeans is the best Open Source java IDE for Enterprise Development and from Apache.

**Comment by wredbeard** [comment]
This would be a great addition to the most non fussy but still comprehensive Java IDE.

**Comment by tzutoo** [comment]
That would be great if this has a NetBeans version.

**Comment by diegofalcinelli** [comment]
https://github.com/TerminalFi/LSP-copilot
https://github.com/intitni/CopilotForXcode

Take a look at these Copilot on Xcode integration projects.

**Comment by n2029-ndensan** [comment]
I want netbeans to support it. Otherwise, I think netbeans will not be used.

**Comment by banifou** [comment]
Yes, copilot plugin for NetBeans please. Otherwise soon nobody will use Netbeans anymore. 
Studies show that the next generation of copilot will increase work efficiency by a factor of 8 or 10!

**Comment by oxido21** [comment]
They  use VS with copilot to develop netBeans, so don't bother anymore :)) 
Personaly, I will start to try to use VS, the only thing that it keeps me with netbeans, it is the posibbility to work with files on remote servers, but if I will find that I can do the same on VS easly, I will stop use netbeans :( ( I realy like netBeans)

**Comment by kevin-hc** [comment]
Sí, complemento de copiloto para NetBeans, por favor. De lo contrario, pronto nadie usará más Netbeans.
¡Los estudios muestran que la próxima generación de copiloto aumentará la eficiencia del trabajo en un factor de 8 o 10!

**Comment by 40yachagi** [comment]
Yes, I am using Netbeans during 15 years, I love it, it is like my office, but if there is no Copilot it becomes unusable :((

**Comment by DeveloperArantec** [comment]
More AI, more Copilot... like professional hunters disappeared, software engineers will, soon... just code for fun!

**Comment by Max84** [comment]
Netbeans can be used for PHP too, not only Java. Copilot can be great with Netbeans :)

**Comment by Hillrunner2008** [comment]
I  started writing a small plugin for chatgpt that some might find useful - https://github.com/Hillrunner2008/netbeans-chatgpt . I certainly don't want anyone to think its comparable to a fully featured co-pilot plugin (fingers crossed that happens), but for those interested, this exist and I'll probably be adding some additional features as my time allows (e.g. some context menu actions from the code editor for refactoring suggestions). At the moment it really isn't much more than a way to stay in netbeans and chat with chatgpt, but for what its worth this is available if you have interest. 

I hesitated to even share it given how basic its feature set is today, but if i see interest i'm more likely to add new features so I'm sharing.

**Reply by Chris2011** [reply]
Thx, will have a look soon to check it out :)

**Reply by banifou** [reply]
I had the same idea but never had the time. Cool, I'll try it.

**Reply by oxido21** [reply]
This is a start, noice! :D

**Reply by banifou** [reply]
Already using it. Works fine so far!

**Reply by manojkumarredbus** [reply]
I just finished integrating open weight models like Llama3.1 and others via Ollama to this plugin. Is it worth releasing?

**Reply by manojkumarredbus** [reply]
Available for locally running Ollama -> https://github.com/manojkumarredbus/netbeans-Ollama
I am yet to finish it properly but works as is with locally hosted Ollama Llama3.1 models.

**Reply by Chris2011** [reply]
hey @manojkumarredbus awesome work. Really. Thx for that :). What about to add those stuff to the chatgpt too? I mean we could also have one more generic named plugin to see that it works with ChatGPT or with other stuff. Both plugins do really the same with different LLMs so please think about to merge this with @Hillrunner2008 plugin to make one big AI based plugin for NetBeans :).

**Reply by Chris2011** [reply]
I see, that you already forked it so it would be good if we can have just one plugin which fits all. Just work together and others, like me, will follow, if I find the time.

**Reply by manojkumarredbus** [reply]
Okay, let me try that, merging should not be that hard.

**Reply by manojkumarredbus** [reply]
@Chris2011 Merge for Ollama and OpenAI on my  fork is done but OpenAI flow needs one level of testing.  Readme is updated appropriately.

**Reply by Hillrunner2008** [reply]
> Okay, let me try that, merging should not be that hard.

I am happy to review a PR @manojkumarredbus and cut a release with your feature properly merged in a way that allows users to choose. I have many ideas and very little time, but I can commit to reviewing your work and contributing to integrating the option in an intuitive way for others to benefit from without needing to download a separate plugin if you want to handle it that way. 

If you want to keep the fork and not bother no worries.

**Comment by luiscla27** [comment]
Any news on this?

**Comment by helderneres** [comment]
> Does Copilot have a Netbeans IDE extenstion? It's my preferred IDE.

Add NetBeans support please! NetBeans is great!

**Comment by lawnmower425** [comment]
We dearly need that - netbeans is a great alternative to phpstorm in the lampp-environment

**Comment by aks2161989** [comment]
Don't let them kill open source! Bring copilot for Netbeans and Eclipse please!

**Reply by EktorOrmaexteaV** [reply]
there's a copilot plug-in in eclipse, mate

**Comment by ermya** [comment]
Yes, copilot plugin for NetBeans please 😢😢
I have a strong preference for NetBeans and I am unwilling to switch to another IDE solely for the sake of accessing Copilot

**Reply by Liandyrf** [reply]
Yes, copilot plugin for NetBeans please

**Comment by oh-jason-palmer** [comment]
Yes, copilot plugin for Netbeans please.

**Comment by Vitexus** [comment]
Dear community, please stop begging and start contributing. 
Those who cannot send the code can contribute financially. 
https://netbeans.apache.org/front/main/participate/

**Reply by ErickPol** [reply]
I would agree if Copilot was an open source project. However, Copilot is a paid, proprietary service offered by Microsoft - in this case I believe it is reasonable to assume that the company itself should take a more active role in developing integrations and plugins.
I find it very hard to see any contribution, especially for free, to a revenue-making product from a large corporation : to my view this is really a question of whether MS desires to minimally collaborate with the same open source community that is providing it, no royalties charged, with the raw materials to train both Copilot and GPT.

**Reply by Vitexus** [reply]
I Already use the copilot with NetBeans everyday. 

**It can be triggered by pressing ALT-TAB**

**Reply by Pha12thom** [reply]
Wow! Gotta try this one out

**Comment by danitb1994** [comment]
Si, me gustaría Copilot para NetBeans

**Comment by quantrpeter** [comment]
Yes, copilot plugin for NetBeans please

**Comment by pmiagree** [comment]
Yes, copilot plugin for NetBeans please.

**Comment by Cibor** [comment]
Yes, copilot plugin for NetBeans please.

**Comment by auffi** [comment]
Yes, copilot plugin for NetBeans please.

**Comment by qmegas** [comment]
Netbeans will be stop used if it won't make steps for AI helpers integration.

**Reply by Vitexus** [reply]
I do not think so ...

**Reply by qmegas** [reply]
> I do not think so ...

In that case, you are not far-sighted.

**Reply by Chris2011** [reply]
NetBeans still exists nowadays without a lot of needed features compared to other IDEs and it will still exists. Also we can create a plugin for NetBeans for example with usage of local llms as a woraround.

**Reply by banifou** [reply]
It's a very good idea actually, a local LLM which can be fine tuned with the own projects. Everything stay local and save. Any ideas which LLMs could be considered that are not too large?

**Reply by Vitexus** [reply]
> > I do not think so ...
> 
> In that case, you are not far-sighted.

I'm using NetBeans from version 6.
When I need some AI specific things then use ALT-TAB switch to Code and then back to NetBeans.

**Reply by banifou** [reply]
I use Netbeans for very long time too, under Linux. What do you mean by switching and ALT-TAB?? Not every body work with windows!

**Reply by Vitexus** [reply]
> I use Netbeans for very long time too, under Linux. What do you mean by switching and ALT-TAB?? Not every body work with windows!

Haha @banifou is funny man!

![image](https://github.com/community/community/assets/2621130/4284343b-2a43-49db-ab61-7f340766a61f)

**Reply by Hillrunner2008** [reply]
Langchain4j has come a long way if someone is willing to write a plugin or expand my little POC (https://github.com/Hillrunner2008/netbeans-chatgpt). I think it would be fantastic to use the existing netbeans APIs to add some context menus for some quick actions on selected text, classes, or whole directories of source code. 

I've had a busy year, but when I stop burning myself out at work i'm planning to add some useful things i would actually use to the plugin. Even better I'd love to see from forks or PRs #opensource

**Reply by adamoutler** [reply]
LLM Pair Programming is an absolute requirement in 2024.

**Reply by manojkumarredbus** [reply]
Based on @Hillrunner2008 's work, here is plugin with open weight models like Llama3.1 and more 
https://github.com/manojkumarredbus/netbeans-Ollama/tree/master
![image](https://github.com/user-attachments/assets/85b36fab-63f4-49b9-a44d-a59c916d2916)

**Reply by TeKett** [reply]
I mean, from Microsoft's perspective, its a good thing that netbeans goes away. They want you to use Visual Studio, and not a competitor. Now, i would do that if it wasn't for the fact that Netbeans is just better imo, i dont program cpp. Microsoft has some stupid ideas like not auto indenting code on new lines because "not enough people want that feature" (did they add this in 2022?).

**Comment by irfina** [comment]
Wow, I really appreciate your guys efforts on this. Thank you.


Sent from Yahoo Mail for iPhone


On Tuesday, August 27, 2024, 1:02 PM, manojkumarredbus ***@***.***> wrote:




Based on @Hillrunner2008 's work, here is plugin with open weight models like Llama3.1 and more
https://github.com/manojkumarredbus/netbeans-Ollama/tree/master
image.png (view on web)

**Comment by jGauravGupta** [comment]
[Jeddict AI Assistant for Apache NetBeans](https://plugins.netbeans.apache.org/catalogue/?id=103) with features:

- Intelligent Code Completion
- Automated Error Fixing
- Smart Variable Naming
- Smart Method Naming
- Code Generation from Comments
- JavaDoc Generation
- Code Refactoring Suggestions
- Contextual Code Suggestions

<img src="https://github.com/jeddict/jeddict.github.io/blob/master/jeddict-ai/resource/0-SmartSuggestions.gif?raw=true" />

**Reply by banifou** [reply]
The AI Assistant is not mentioned on your homepage https://jeddict.github.io/index.html.
Anyway, it looks impressive. I'll give it a try.

**Reply by jGauravGupta** [reply]
Please refer to the [https://jeddict.github.io/page.html?l=tutorial/AI](https://jeddict.github.io/page.html?l=tutorial/AI) page for the documentation. Additional updates will be made to other pages on the [Jeddict website](https://jeddict.github.io/) soon.

**Reply by Chris2011** [reply]
This is great but there is a big but, it is just for Java, not for the whole IDE with all available languages. What we can do here is to make this available for all languages, this should be possible like VS Code does with copilot

[![image](https://github.com/user-attachments/assets/00ae6836-e907-4453-84b0-132ba496b8c2)
](https://learn.microsoft.com/en-us/visualstudio/ide/media/vs-2022/github-copilot-extension-example.gif?view=vs-2017&viewFallbackFrom=vs-2022)

Because NetBaens knows virtual-text/readonly text and this can be generated from the response from the AI. I will also have a look for this for each language.

**Reply by jGauravGupta** [reply]
Thanks for the feedback! Expanding support to all file types is already planned, and work is in progress. Though the tool is evolving to support a wider range of file types within the NetBeans ecosystem, its core focus will remain on delivering specialized features tailored specifically for Java, Jakarta EE and Web developers, ensuring deep integration and optimal functionality. See the following demo of the upcoming version. Stay tuned for more updates!

![HTML](https://github.com/user-attachments/assets/4747c972-6fa4-42b2-8e89-f50e36c6174d)

**Reply by jShiwaniGupta** [reply]
The latest v1.5 of the JEDDICT AI plugin, available [here](https://plugins.netbeans.apache.org/catalogue/download/555/jeddict-ai-1.5.nbm), now extends support to multiple languages, not just Java.

**Reply by banifou** [reply]
I entered a wrong API key. Now, a window pops up every time and shows me that the API key is wrong. How can I enter a new one?!
I just submitted an issue report: https://github.com/jeddict/jeddict/issues/554

**Reply by jShiwaniGupta** [reply]
Hi @banifou, The issue has been fixed and released in Jeddict AI Assistant v1.6: https://plugins.netbeans.apache.org/catalogue/download/556/jeddict-ai-1.6.nbm

**Reply by banifou** [reply]
Wow, that was very fast! Not everything has been fixed but now I can start testing the assistant. Thanks
I'm also already preparing a list of improvements...

**Reply by banifou** [reply]
Unfortunately, the assistant is still very buggy!
It changes far too much! Even if you only have the javadocs of a method created, the entire file is rewritten. All empty lines are deleted. Indentations and comments are changed. The entire file looks different. If the file is under version control, almost everything is marked blue (for changed). This is very impractical. You no longer know what I have written myself and what the assistant has changed!

Only the javadocs should be added in the right place. Nothing else! The javadocs were also completely wrong.

I have not tested the other features.

Suddenly the assistant stops working! No idea why.

Also, I would like to know, for each function/feature, what is sent to OpenAi? Is only the code of the method or the whole code of the Java class sent to OpenAi?

**Reply by jShiwaniGupta** [reply]
Hi @banifou,

Thank you for the valuable feedback! Concerns regarding the current behavior of the assistant, particularly the rewriting of entire files when only Javadocs need to be added, are noted. This issue is recognized, and improvements are in progress to ensure that only the relevant snippets are updated without altering the rest of the code. This issue is only related to Javadoc generation from Hints.

In the meantime, trying the code completion feature by using Ctrl+Space is recommended, as it updates only the specific snippets intended for modification.

For the tool stopping unexpectedly, please create a ticket and share the NetBeans log for further investigation.

> Also, I would like to know, for each function/feature, what is sent to OpenAi? Is only the code of the method or the whole code of the Java class sent to OpenAi?

Currently Java Source is shared, also its scope can be increased via settings in Options. 

Your input is crucial for refining the assistant. Thank you for your patience and for trying out the tool!

**Reply by banifou** [reply]
I forgot to mention, after the changes of the assistant, the Save, Redo and Undo buttons hasn't been activated. It must be possible to redo the changes at any time!
For now I stopped using the assistant until substantial improvements have been made. 
I'm curious to see what other users are saying.
Good luck

**Reply by Chris2011** [reply]
hey @banifou great that you are testing it, please focus your thoughts, bugs and suggestions, where it belongs. To the jeddict AI assitant repo. Also when you cretae tickets, don't create a batch ticket for differnet stuff. One ticket per bug and also for feature request. Thx :)

**Reply by jShiwaniGupta** [reply]
Hi @banifou,

Thank you for the valuable feedback regarding the Redo/Undo functionality and the issues with the entire file being rewritten. These issues have been addressed and released in Jeddict AI Assistant v1.7. The latest version can be downloaded here: [Jeddict AI Assistant v1.7](https://plugins.netbeans.apache.org/catalogue/download/557/jeddict-ai-1.7.nbm).

Further observations or issues are encouraged, as such input is crucial for refining the assistant. Continuous improvements are the goal.

Thank you for your support!

**Reply by banifou** [reply]
@jShiwaniGupta - you are welcome.

**Comment by irfina** [comment]
Wow, very good


Sent from Yahoo Mail for iPhone


On Tuesday, September 24, 2024, 7:23 PM, Gaurav Gupta ***@***.***> wrote:




Jeddict AI Assistant for Apache NetBeans

**Comment by CarlosXUSAC** [comment]
Yo regrese a la Universidad tras 25 años y conocia de Programacion Estructurada, y el cambio a POO me freno mucho al punto de pensar en retirarme de nueve, sin enbargo el uso de Copilot me ha ayudado a generar y corregir codigo, ya que los plazos de los Proyectos son muy cortos y es frustrante dedicarle todo mi tiempo y al final no lograr terminarlo, pero todo cambio pues esta herramienta me ayuda a que el codigo me fluya y logre encontrar mis errores, gracias, Uso IntelliJ pero muchas veces me obligan a usar Nerbeans. Es un cambio muy drastico el no contar con ese soporte.

**Comment by lkishalmi** [comment]
GitHub Copilot, as of now, does not provide a public, general-purpose API for direct integration into third-party applications or plugins (such as desktop apps). Copilot is mainly available through:

- GitHub Copilot extension for Visual Studio Code, JetBrains IDEs, and Neovim
- GitHub Copilot Chat in supported IDEs
- GitHub API for Copilot (for organizations, usage analytics, and admin controls, but not for direct code completion or chat)

**If you want to integrate Copilot-like capabilities (AI code suggestions) in your desktop application, here are your options:**

### 1. Official GitHub Copilot API Documentation
There is **no public, general Copilot API documentation** provided by GitHub for third-party integration.  
- [GitHub Copilot Docs (User & Admin)](https://docs.github.com/en/copilot)
- [Copilot for Business/Enterprise API](https://docs.github.com/en/rest/copilot) (analytics/admin only)

### 2. Alternatives
- **Copilot Chat and Suggestions** are not exposed as a public API.
- **Copilot Enterprise** can be integrated with internal documentation and code search, but not as an API for external apps.

### 3. If you want AI code completion in your app:
- You will need to use a different API, such as [OpenAI's API](https://platform.openai.com/docs/guides/code) (which Copilot is based on).
- Or, build a plugin that interfaces with the Copilot VS Code extension, but this is not officially supported and may violate Copilot's terms of service.

### 4. Reference Docs
- [Copilot REST API (for admins/analytics)](https://docs.github.com/en/rest/copilot)
- [OpenAI API for code](https://platform.openai.com/docs/guides/code)
- [GitHub Copilot Terms of Service](https://docs.github.com/en/site-policy/github-terms/github-copilot-terms-of-use)

---

**Summary:**  
There is not a public Copilot API you can use in your desktop application. You can review the [Copilot documentation](https://docs.github.com/en/copilot) for available features, but for plugin development, you would need to use a general code AI API like OpenAI's.

If you have a specific use case or want to target a particular IDE, let me know and I can provide more targeted integration guidance.

**Reply by lkishalmi** [reply]
Well, I've just asked Copilot, how to start such a work.
I hope that summary speaks for itself...

**Reply by Chris2011** [reply]
Maybe now everything changes:

https://medium.com/@servifyspheresolutions/microsoft-just-open-sourced-github-copilot-under-mit-license-and-it-changes-everything-for-2673536f2cbb

**Reply by Chris2011** [reply]
So we can have a look into the code now. Maybe the output is the newest that they gave.

**Reply by PavelTurk** [reply]
@Chris2011 Could you let us know if there has been any progress on this issue?

**Reply by Chris2011** [reply]
Hey @PavelTurk I don't know what you expect here, it was just a comment, that it could be possible. Atm there is no progress from my side.

**Reply by PavelTurk** [reply]
@Chris2011 For some reason, I thought you were part of the NetBeans team. My mistake, sorry.

**Reply by Chris2011** [reply]
@PavelTurk  Well there is no real netbeans team anymore as of NetBeans is part of Apache and completely open source. So we all are a part of NetBeans when we want to. But this is not the right place to tell you this. If you have questions, please send me a mail, you can find it on my github profile.

**Comment by EktorOrmaexteaV** [comment]
NetBeans is amazing, specially for GUI developing in Java and they'll get even better if they allow the integration with AI

so yes, copilot plugin for NetBeans please

**Comment by econopapi** [comment]
yes, copilot plugin for NetBeans please

**Comment by priyanahata** [comment]
How about this one? https://www.anahata.uno/

<img width="1201" height="897" alt="image" src="https://github.com/user-attachments/assets/b4b995c7-9f86-4a14-a508-e913e376b825" />

**Comment by irfina** [comment]
Looks interesting, will give it a try. Thanks for sharing.


Sent from Yahoo Mail for iPad


On Saturday, January 31, 2026, 8:33 AM, Priyadarshi ***@***.***> wrote:




How about this one? https://www.anahata.uno/
image.png (view on web)

**Reply by priyanahata** [reply]
It will do anything and everything for you, not just code on netbeans, it can turn down your tv too

**Reply by priyanahata** [reply]
@irfina

**Reply by Chris2011** [reply]
@irfina please, use it with caution, it seems that this tool is somehow scam.

**Comment by priyanahata** [comment]
@irfina dont listen to @Chris2011 . I is Not "somehow" i am the number one comitter, there is no "somehow" it is either a scam or it "is" a scam
